### PR TITLE
[Ray DAG] Add `execute()` interface to take user inputs with InputNode.

### PR DESCRIPTION
**Pull request by jiaodong**
## Diff Summary

Current implementation of DAGNode pre-bind inputs and the signature of `def execute(self)` doesn't take user input yet. This PR extends the interface to take user input, mark DAG entrypoint methods as first stop of all user requests in a DAG. It's needed to unblock next step serve pipeline implementation to serve user requests.

Closes #22196 #22197

Notable changes:
- Added a `DAG_ENTRY_POINT` flag in ray dag API to annotate DAG entrypoint functions. Function or class method only. All marked functions will receive identical input from user as first layer of DAG. 
- Changed implementations of ClassNode and FunctionNode accordingly to handle different execution for a node marked as entrypoint or not.
- Added a `kwargs_to_resolve` kwarg in the interface of `DAGNode` to handle args that sub-classes need to use to resolve it's implementation without exposing all terms to parent class level.
  - This is particularly important for ClassMethodNode binding, so we can have implementations to track method name, parent ClassNode as well as previous class method call without existiting 
  - Changed implementation of `_copy()` to handle execution of `kwargs_to_resolve`.
  - Changed implementation of `_apply_and_replace_all_child_nodes()` to fetch DAGNode type in `kwargs_to_resolve`.
- Added pretty printed lines for `kwargs_to_resolve`

## Test Plan

All existing and added unit tests passed with only one change needed to migrate a complex nested args test to new API in test. 

This is a sample multi-model ensemble DAG:

```
m1 = Model._bind(2)
m2 = Model._bind(3)

m1_output = m1.forward._bind(ray_dag.DAG_ENTRY_POINT)
m2_output = m2.forward._bind(ray_dag.DAG_ENTRY_POINT)

dag = combine._bind(m1_output, m2_output)
```

This is the pretty-printed DAG structure: 

```
 ray/experimental/dag/tests/test_dag_entrypoint.py::test_multi_class_method_entrypoint ✓                                                                                                                                        50% █████     (FunctionNode)(
    body=<function test_multi_class_method_entrypoint.<locals>.combine at 0x7f8b98e6f040>
    args=[
        (ClassMethodNode)(
            body=forward()
            args=[
                __dag_entry_point, 
            ]
            kwargs={}
            options={}
            kwargs_to_resolve={
                method_name: forward
                parent_class_node:    
                    (ClassNode)(
                        body=<class 'test_dag_entrypoint.modify_class.<locals>.Class'>
                        args=[
                            2, 
                        ]
                        kwargs={}
                        options={}
                        kwargs_to_resolve={}
                    )
                prev_class_method_call: None
            }
        )
        (ClassMethodNode)(
            body=forward()
            args=[
                __dag_entry_point, 
            ]
            kwargs={}
            options={}
            kwargs_to_resolve={
                method_name: forward
                parent_class_node:    
                    (ClassNode)(
                        body=<class 'test_dag_entrypoint.modify_class.<locals>.Class'>
                        args=[
                            3, 
                        ]
                        kwargs={}
                        options={}
                        kwargs_to_resolve={}
                    )
                prev_class_method_call: None
            }
        )
    ]
    kwargs={}
    options={}
    kwargs_to_resolve={}
)
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/dag/__init__.py [hunk 1]
```diff
@@ -1,6 +1,13 @@
 from ray.experimental.dag.dag_node import DAGNode
 from ray.experimental.dag.function_node import FunctionNode
 from ray.experimental.dag.class_node import ClassNode, ClassMethodNode
+from ray.experimental.dag.constants import DAG_ENTRY_POINT
```
Comment (ericl) [inline python/ray/experimental/dag/__init__.py:4]:
```suggestion
from ray.experimental.dag.constants import INPUT
```

How about shortening this? I think dag.INPUT (or input_args, etc.) is pretty clear. Entrypoint sounds a bit esoteric.

**Review by ericl — COMMENTED** [review]
Couple questions:
- Do we intend to support using ENTRY_POINT in conjunction with other params? For example, `f._bind(ENTRY_POINT, 1, 2, 3)`?
- What about extracting specific kwargs of entrypoint, like `f._bind(ENTRY_POINT["foo"])`?

It's fine if we don't, just want to clarify if this is a TODO or explicitly not supported. We should also raise error messages if ENTRY_POINT is used in an unsupported way.

**Comment by jiaodong** [comment]
> Do we intend to support using ENTRY_POINT in conjunction with other params? For example, f._bind(ENTRY_POINT, 1, 2, 3)?

I think we shouldn't and should assert failures in this case. ENTRY_POINT essentially makes all args & kwargs from user as input to the function. I can address this in this PR. 

> What about extracting specific kwargs of entrypoint, like f._bind(ENTRY_POINT["foo"])?

I don't quite see why we want this assuming ENTRY_POINT can only be used as the only input in binding. Looks like it makes things a bit more complicated than needed tho, what use cases it will enable ?

File: python/ray/experimental/dag/__init__.py (hunk 1 above)
Comment (jiaodong) [reply python/ray/experimental/dag/__init__.py:4]: sure i think this is simpler too

**Comment by ericl** [comment]
Ok, let's leave a todo to revisit this prior to moving out of experimental then. I think this is a little limiting, but agree the use cases are not super clear for the advanced cases.

File: python/ray/experimental/dag/class_node.py [hunk 2]
```diff
@@ -63,57 +89,72 @@ def options(self, **options):
...
-        method_name: str,
         method_args: Tuple[Any],
         method_kwargs: Dict[str, Any],
         method_options: Dict[str, Any],
+        kwargs_to_resolve: Dict[str, Any],
     ):
-        self._method_name: str = method_name
+        # TODO: (jiaodong) revisit constraints on dag INPUT before moving out
+        # of experimental folder
+        if ray_dag.INPUT in method_args and len(method_args) > 1:
```
Comment (ericl) [inline python/ray/experimental/dag/class_node.py:104]: Can we make this a type that overrides `__reduce__` so it can't be accidentally captured even nested in args?

File: python/ray/experimental/dag/constants.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+# In this pipeline, each user input is broadcasted to both A.forward and
+# B.forward as first stop of the DAG, and authored like
+
+# a = A.forward.bind(ray.dag.INPUT)
+# b = B.forward.bind(ray.dag.INPUT)
+# dag = ensemble.bind(a, b)
+
+# dag.execute(user_input) --> broadcast to a and b
+INPUT = "__INPUT"
```
Comment (ericl) [inline python/ray/experimental/dag/constants.py:19]: Make it a non-serializable value for safety.

File: python/ray/experimental/dag/class_node.py [hunk 4]
```diff
@@ -47,13 +72,14 @@ def __init__(self, actor: ClassNode, method_name: str):
         self._options = {}
 
     def _bind(self, *args, **kwargs):
+        kwargs_to_resolve = {
```
Comment (ericl) [inline python/ray/experimental/dag/class_node.py:75]: Instead of calling this `kwargs_to_resolve` can we call it `other_args_to_resolve`

File: python/ray/experimental/dag/class_node.py [hunk 5]
```diff
@@ -47,13 +72,14 @@ def __init__(self, actor: ClassNode, method_name: str):
...
+            "method_name": self._method_name,
```
Comment (ericl) [inline python/ray/experimental/dag/class_node.py:76]: I don't think method name needs to be resolved.

File: python/ray/experimental/dag/tests/test_dag_input.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        def __init__(self, val):
+            self.val = val
+
+        def get(self):
+            return self.val
+
+    with pytest.raises(
+        ValueError, match="INPUT cannot be used as ClassNode args"
+    ):
+        Actor._bind(ray_dag.INPUT)
```
Comment (ericl) [inline python/ray/experimental/dag/tests/test_dag_input.py:101]: Actually, why can't it be used to bind here?

File: python/ray/experimental/dag/tests/test_dag_input.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    with pytest.raises(
+        ValueError,
+        match="dag.INPUT cannot be used in conjunction with other args",
+    ):
+        f._bind(ray_dag.INPUT, 1)
+
+    with pytest.raises(
+        ValueError, match="dag.INPUT cannot be used as a kwarg value"
+    ):
+        f._bind(1, kwarg=ray_dag.INPUT)
```
Comment (ericl) [inline python/ray/experimental/dag/tests/test_dag_input.py:118]: Add a test for `f._bind([[{"nested": ray_dag.INPUT}]])`

File: python/ray/experimental/dag/tests/test_dag_input.py (hunk 6 above)
Comment (edoakes) [reply python/ray/experimental/dag/tests/test_dag_input.py:101]: Don't think this makes sense; we want to explicitly separate the "instantiate the DAG" and "execute the DAG on an input" steps (that's the purpose of having a static DAG definition based on my understanding)

File: python/ray/experimental/dag/class_node.py [hunk 8]
```diff
@@ -1,26 +1,51 @@
...
+        cls: type,
+        cls_args,
+        cls_kwargs,
+        cls_options,
+        kwargs_to_resolve=None,
+    ):
+        if ray_dag.INPUT in cls_args or ray_dag.INPUT in cls_kwargs.values():
+            raise ValueError(
+                "dag.INPUT cannot be used as ClassNode args. "
+                "Please bind to function or class method only."
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:21]:
```suggestion
                "dag.INPUT cannot be used as ClassNode args because it isn't available at class construction time. "
                "Please bind to function or class method only."
```
users will inevitably be confused as to why :)

File: python/ray/experimental/dag/class_node.py [hunk 9]
```diff
@@ -1,26 +1,51 @@
...
+            )
         self._body = cls
         self._last_call: Optional["ClassMethodNode"] = None
-        DAGNode.__init__(self, cls_args, cls_kwargs, cls_options)
+        ray_dag.DAGNode.__init__(
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:25]: hmm why are we doing this wonky manual call to `DAGNode.__init__`? Are you trying to call the parent constructor? If so, do `super().__init__()`

File: python/ray/experimental/dag/class_node.py [hunk 10]
```diff
@@ -1,26 +1,51 @@
...
+            cls_options,
+            kwargs_to_resolve=kwargs_to_resolve,
+        )
 
     def _copy_impl(
         self,
         new_args: List[Any],
         new_kwargs: Dict[str, Any],
         new_options: Dict[str, Any],
+        new_kwargs_to_resolve: Dict[str, Any],
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:38]: can you please add a docstring describing what each of these is? hard to parse w/o context

File: python/ray/experimental/dag/class_node.py [hunk 11]
```diff
@@ -47,13 +72,14 @@ def __init__(self, actor: ClassNode, method_name: str):
         self._options = {}
 
     def _bind(self, *args, **kwargs):
+        kwargs_to_resolve = {
+            "method_name": self._method_name,
+            "parent_class_node": self._actor,
+            "prev_class_method_call": self._actor._last_call,
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:78]: what's this used for?

File: python/ray/experimental/dag/class_node.py [hunk 12]
```diff
@@ -1,26 +1,51 @@
 import ray
-from ray.experimental.dag.dag_node import DAGNode
+import ray.experimental.dag as ray_dag
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:2]:
```suggestion
from ray.experimental.dag import DAGNode, INPUT
```
better to do specific imports

File: python/ray/experimental/dag/class_node.py (hunk 2 above)
Comment (edoakes) [reply python/ray/experimental/dag/class_node.py:104]: good idea. clarifying question though -- won't we detect it in nested args due to the cloudpickling magic to find & replace nodes?

File: python/ray/experimental/dag/class_node.py [hunk 13]
```diff
@@ -63,57 +89,72 @@ def options(self, **options):
...
     ):
-        self._method_name: str = method_name
+        # TODO: (jiaodong) revisit constraints on dag INPUT before moving out
+        # of experimental folder
+        if ray_dag.INPUT in method_args and len(method_args) > 1:
+            raise ValueError(
+                "dag.INPUT cannot be used in conjunction with other args. "
+                "Please bind with dag.INPUT as only input."
+            )
+        if ray_dag.INPUT in method_kwargs.values():
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:109]: also check that it isn't in method_options

File: python/ray/experimental/dag/class_node.py [hunk 14]
```diff
@@ -63,57 +89,72 @@ def options(self, **options):
...
+            raise ValueError(
+                "dag.INPUT cannot be used as a kwarg value. "
+                "Please bind with dag.INPUT as only input."
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:112]:
```suggestion
                "dag.INPUT cannot be used as a kwarg value. "
                "Please bind with dag.INPUT as the only input."
```

File: python/ray/experimental/dag/class_node.py [hunk 15]
```diff
@@ -63,57 +89,72 @@ def options(self, **options):
...
         method_options: Dict[str, Any],
+        kwargs_to_resolve: Dict[str, Any],
     ):
-        self._method_name: str = method_name
+        # TODO: (jiaodong) revisit constraints on dag INPUT before moving out
+        # of experimental folder
+        if ray_dag.INPUT in method_args and len(method_args) > 1:
+            raise ValueError(
+                "dag.INPUT cannot be used in conjunction with other args. "
+                "Please bind with dag.INPUT as only input."
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:107]:
```suggestion
                "Please bind with dag.INPUT as the only input."
```

File: python/ray/experimental/dag/class_node.py [hunk 16]
```diff
@@ -63,57 +89,72 @@ def options(self, **options):
...
-            return (
-                getattr(actor_handle, self._method_name)
-                .options(**self._bound_options)
-                .remote(
-                    *self._bound_args[2:],
-                    **self._bound_kwargs,
-                )
-            )
+        method_body = getattr(self._parent_class_node, self._method_name)
+        if len(self._bound_args) == 1 and self._bound_args[0] == ray_dag.INPUT:
```
Comment (edoakes) [inline python/ray/experimental/dag/class_node.py:152]: move this to a utility method?

Comment (edoakes) [reply python/ray/experimental/dag/class_node.py:152]: `self._is_dag_entrypoint`

File: python/ray/experimental/dag/constants.py (hunk 3 above)
Comment (edoakes) [reply python/ray/experimental/dag/constants.py:19]: can make it an ENUM and override `__reduce__`

**Review by edoakes — COMMENTED** [review]
Detailed comments inline. A few higher-level questions:

- `dag.execute()` looks nice for testing, but if I wanted to actually use this thing I'd want to separate instantiate and then execute it. Should we do `dag.remote()` and have a "dag handle" to match the rest of the Ray API here? Thoughts @ericl?
- Do we intend to always require `INPUT` as the sole argument to input nodes? It would be a pretty strange use case to bind other arguments to it and there's no loss of generality because you can bind them to constructor, just clarifying.

File: python/ray/experimental/dag/class_node.py (hunk 2 above)
Comment (ericl) [reply python/ray/experimental/dag/class_node.py:104]: We'd have to explicitly add a case looking for that in the cloudpickling scanner then. Right now it only intercepts DAGNode types.

File: python/ray/experimental/dag/tests/test_dag_input.py (hunk 6 above)
Comment (ericl) [reply python/ray/experimental/dag/tests/test_dag_input.py:101]: I guess it makes sense if we plan to add the separation in the future.

**Comment by ericl** [comment]
> dag.execute() looks nice for testing, but if I wanted to actually use this thing I'd want to separate instantiate and then execute it. Should we do dag.remote() and have a "dag handle" to match the rest of the Ray API here? Thoughts @ericl?

That would only be a performance optimization right? Because right now each execution will create fresh actors and destroy them after execution.

Besides separating the stages, we could also achieve the same optimization with a flag like `dag.options(reuse_actors=True)`.

> Do we intend to always require INPUT as the sole argument to input nodes? It would be a pretty strange use case to bind other arguments to it and there's no loss of generality because you can bind them to constructor, just clarifying.

I think we should allow it. For example, even in this basic example it is already using INPUT mixed with other arguments on line 87: https://gist.github.com/ericl/dc73a7b19b6c1e13d6f2f8e8a72e62f3

This could be implemented in a separate PR though.

**Comment by ericl** [comment]
Another possibility is to require explicit expansion of input as args and kwargs, for example:

```
with ray.dag.builder() as input:
    f.remote(input[0], input[1])
    f.remote(a=input["a"] b=input["b"])
    f.remote(*input.args, **input.kwargs)
```

File: python/ray/experimental/dag/class_node.py (hunk 2 above)
Comment (jiaodong) [reply python/ray/experimental/dag/class_node.py:104]: sg, this is more robust solution.

File: python/ray/experimental/dag/tests/test_dag_input.py (hunk 6 above)
Comment (jiaodong) [reply python/ray/experimental/dag/tests/test_dag_input.py:101]: I'm treating `ray_dag.INPUT` as user request sent to a pre-built, static DAG to execute, rather than a handy way to dynamically redefine a class constructor. We have a number of choices that can be considered "limiting" at first glance but probably better off guard-railed to begin with, such as enforcing JSON-serializable objects only in binding.

File: python/ray/experimental/dag/class_node.py (hunk 11 above)
Comment (jiaodong) [reply python/ray/experimental/dag/class_node.py:78]: it's to build lineage of class method calls in ray dag, so a ClassNode as parent can spawn multiple ClassMethodNodes in a chain

File: python/ray/experimental/dag/class_node.py (hunk 12 above)
Comment (jiaodong) [reply python/ray/experimental/dag/class_node.py:2]: that's what i had but it will lead to circular imports warning so i just left public imports at dag __init__ level and just import the module here

File: python/ray/experimental/dag/class_node.py (hunk 16 above)
Comment (jiaodong) [reply python/ray/experimental/dag/class_node.py:152]: sg, will add

**Comment by jiaodong** [comment]
> Another possibility is to require explicit expansion of input as args and kwargs, for example:
> 
> ```
> with ray.dag.builder() as input:
>     f.remote(input[0], input[1])
>     f.remote(a=input["a"] b=input["b"])
>     f.remote(*input.args, **input.kwargs)
> ```
> 
> Slicing the `input` object could generate a DAGNode that returns the specified part of the input, no special handling needed, and it would also support nested inputs like this automatically:
> 
> ```
> f.remote(data={"foo": input["foo"]})
> ```

I think introducing an `InputNode(DAGNode)` class makes sense to handle nested args, and it integrates well with current implementations.

I don't think we should encourage user to do dynamic or implicit behavior in DAG binding however, like disallowing `f.remote(*input.args, **input.kwargs)`. This could make serve artifact generation harder since whole DAG and bound args are not fully resolved until we run the DAG with an actual user input that's subject to change.

**Comment by jiaodong** [comment]
So for next step let me update this PR with
1) Move `dag.INPUT` to a `InputNode(DAGNode)` class type, start simple as if it's a dataclass where everything is explicitly given.
2) Address comments regarding tests, docstring, variable names, etc.
3) Leave the restriction of no `dag.INPUT` in actor constructor binding as-is with comments, we can revisit this later.

**Comment by ericl** [comment]
> Leave the restriction of no dag.INPUT in actor constructor binding as-is with comments, we can revisit this later.

Could we just raise an error in Serve only? It seems fine for general DAGs. Serve is doing other things like checking JSON serialization as well.

**Comment by jiaodong** [comment]
> > Leave the restriction of no dag.INPUT in actor constructor binding as-is with comments, we can revisit this later.
> 
> Could we just raise an error in Serve only? It seems fine for general DAGs. Serve is doing other things like checking JSON serialization as well.

Sure that makes sense too. IIUC at ray core level, ray DAG is dynamic in nature and we want to optimize for expressiveness; In serve, a DAG should be static with limitations to ensure 1) DAG is statically defined 2) All bound args are JSON serializable 3) other perf / operational changes.

**Comment by edoakes** [comment]
I'd prefer we disallow the `INPUT` node being passed to actor constructor nodes entirely because it breaks the semantics to me (the nodes of the graph are now dynamic based on the input, no longer static), but if you feel strongly @ericl I don't mind asserting it in Serve only.

Also @ericl maybe we should add the explicit instantiate then call vs. `cache_actors` flag to our request for feedback? Seems like mostly a "taste" thing.

**Comment by ericl** [comment]
Sure, I think banning it for now is fine since we don't even have an instantiate API (and currently no use case really requires it).

File: python/ray/experimental/dag/class_node.py (hunk 13 above)
Comment (jiaodong) [reply python/ray/experimental/dag/class_node.py]: we're using pickling method now that skips tracking DAGNode's `options`. I can do this in a follow-up at DAGNode level so that `options()` should contain no object of type DAGNode at all that will take care of this case.

File: python/ray/experimental/dag/class_node.py [hunk 17]
```diff
@@ -1,37 +1,60 @@
...
-        DAGNode.__init__(self, cls_args, cls_kwargs, cls_options)
+        super().__init__(
+            cls_args,
+            cls_kwargs,
+            cls_options,
+            other_args_to_resolve=other_args_to_resolve,
+        )
+
+        if self._contain_input_node():
+            raise ValueError(
```
Comment (ericl) [inline python/ray/experimental/dag/class_node.py:29]: "InputNodes cannot be used as arguments to actor constructors, only task and actor method calls."

File: python/ray/experimental/dag/class_node.py [hunk 18]
```diff
@@ -68,52 +95,71 @@ class ClassMethodNode(DAGNode):
...
+            if (
+                len(self.get_args()) != 1
+                or not isinstance(self.get_args()[0], InputNode)
+                or self.get_kwargs() != {}
+            ):
+                raise ValueError(
+                    "InputNode marks the entrypoint of user request to the "
+                    "DAG, please ensure InputNode is the only input to a "
+                    "ClassMethodNode, and NOT used in conjunction with, or "
+                    "nested within other args or kwargs."
```
Comment (ericl) [inline python/ray/experimental/dag/class_node.py:140]: Isn't the point of InputNode to allow mixed input? I think you can just remove this.

File: python/ray/experimental/dag/dag_node.py [hunk 19]
```diff
@@ -166,6 +198,16 @@ def __call__(self, node):
             )
         )
 
+    def _contain_input_node(self) -> bool:
```
Comment (ericl) [inline python/ray/experimental/dag/dag_node.py:201]:
```suggestion
    def _contains_input_node(self) -> bool:
```

File: python/ray/experimental/dag/dag_node.py [hunk 20]
```diff
@@ -166,6 +198,16 @@ def __call__(self, node):
...
+        """Check if InputNode is used in children DAGNodes with current node
+        as the root.
+        """
+        children_dag_nodes = self._get_all_child_nodes()
+        for child in children_dag_nodes:
+            if isinstance(child, ray_dag.InputNode):
+                return True
+        return False
```
Comment (ericl) [inline python/ray/experimental/dag/dag_node.py:209]: Since this is only used in ActorNode, you can move it into that subclass.

File: python/ray/experimental/dag/function_node.py [hunk 21]
```diff
@@ -2,33 +2,69 @@
...
+        )
+        # TODO: (jiaodong) Disallow binding other args if InputNode is used.
+        # revisit this constraint before moving out of experimental folder
+        has_input_node = self._contain_input_node()
+        if has_input_node:
+            # Invalid usecases:
+            # f.bind(InputNode(), 1, 2)
+            # f.bind(1, 2, key=InputNode())
+            # f.bind({"nested": InputNode()})
+            # f.bind(InputNode(), key=123)
```
Comment (ericl) [inline python/ray/experimental/dag/function_node.py:35]: Let's remove this.

File: python/ray/experimental/dag/input_node.py [hunk 22]
```diff
@@ -0,0 +1,47 @@
+from typing import Any, Dict, List
+
+from ray.experimental.dag import DAGNode
+
+
+class InputNode(DAGNode):
```
Comment (ericl) [inline python/ray/experimental/dag/input_node.py:6]: I think we can go ahead and implement the full version, the code is something like this:

```
class InputNode:
    def __init__(self, fetch):
        self.fetch = fetch

    def _execute_impl(self, *args, **kwargs):
        if self.fetch == "__args__":
            return args
        elif self.fetch == "__kwargs__":
            return kwargs
        elif isinstance(self.fetch, int):
            return args[self.fetch]
        else:
            return kwargs[self.fetch]


class InputNodeFactory:
    args = InputNode("__args__")
    kwargs = InputNode("__kwargs__")

    def __getitem__(self, fetch):
        return InputNode(fetch)


## Use like:
# dag.input.args
# dag.input.kwargs
# dag.input[0]
# dag.input["foo"]
input = InputNodeFactory()
```

**Review by ericl — COMMENTED** [review]
Seems easier just to implement the full InputNode--- I attached a snippet that does most of it.

File: python/ray/experimental/dag/class_node.py (hunk 18 above)
Comment (jiaodong) [reply python/ray/experimental/dag/class_node.py:140]: yep with InputNode we can add more flexibility now

File: python/ray/experimental/dag/input_node.py (hunk 22 above)
Comment (jiaodong) [reply python/ray/experimental/dag/input_node.py:6]: hmm is this kinda making InputNodeFactory a singleton that produces args / kwargs accessor, how about we leave it to the next PR to properly set it up with more tests ? So for today I can get more serve pipeline side of work done for the demo.

**Comment by edoakes** [comment]
@ericl this PR is quite long already and tests are green w/ the very dumb version on `InputNode` -- let's merge this so jiao can also make progress on the serve integration in parallel.

@jiaodong let's follow up with a PR to implement the `InputNode` as eric described above.

**Review by edoakes — APPROVED** [review]

**Comment by jiaodong** [comment]
created: https://github.com/ray-project/ray/issues/22255
